### PR TITLE
#167749954 Implement Password Reset Feature

### PR DESCRIPTION
**Pull request by daviddamilola**
#### What does this PR do?
 Add password reset feature
#### Description of Task to be completed?
* add test for password reset endpoint
* add password reset endpoint routes.
* add update Password service.
* add verify email and send email Helper.
* add password validation user schema.
* update documentation.
#### How should this be manually tested?
* `git clone` the branch
* set the environment variables
* run postman
* send a `POST` request to ` /api/v1/users/reset`
* send a `PATCH`  request containing the new password to the link in the email
#### Any background context you want to provide?
`nodemailer` was used to send the email.
#### What are the relevant pivotal tracker stories?
[#167749954](https://www.pivotaltracker.com/story/show/167749954)

### CODE REVIEW

**Review by ldonjibson — APPROVED** [review]

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,11 @@ before_script:
 node_js:
 - 'stable'
 
+env:
```
Comment (hustlaviola) [inline .travis.yml]: You can remove the environment variables, it has been added on Travis

File: src/config/swagger.json [hunk 2]
```diff
@@ -80,10 +193,19 @@
...
           "items": {
             "properties": {
+              "token": {
+                "type": "string"
+              },
               "id": {
                 "type": "integer",
                 "format": "int64"
               },
+              "firstname": {
```
Comment (hustlaviola) [inline src/config/swagger.json]:
```suggestion
              "firstName": {
```

File: src/config/swagger.json [hunk 3]
```diff
@@ -80,10 +193,19 @@
...
+                "type": "string"
+              },
+              "lastname": {
```
Comment (hustlaviola) [inline src/config/swagger.json]:
```suggestion
              "lastName": {
```

File: src/controllers/UserController.js [hunk 4]
```diff
@@ -39,4 +41,41 @@ export default class UserController {
...
+
+  static async updatePassword(req, res) {
+    const token = req.params.token;
+    const data = Helper.verifyToken(token).payload;
+    const hasExpired = Date.now() > data.expiryDate;
+    if (hasExpired) {
+      Responses.setError(400, 'this address link has expired')
+      return Responses.send(res);
+    }
+    if (req.method == 'GET') {
```
Comment (hustlaviola) [inline src/controllers/UserController.js]:
```suggestion
    if (req.method === 'GET') {
```

File: src/database/models/user.js [hunk 5]
```diff
@@ -19,4 +20,4 @@ export default (sequelize, DataTypes) => {
     },
   }, {});
   return User;
-};
+};
```
Comment (hustlaviola) [inline src/database/models/user.js]: You can always add newlines at the end of a file

**Review by Naimaoye — CHANGES_REQUESTED** [review]
Good job David. I think you need to add new lines at the end of some of your files, please check through

**Review by hustlaviola — COMMENTED** [review]
A lot of files does not have newlines at the end. Consider adding them. You can also rebase off `ft-JWT-auth-167749949` to implement some changes on your branch. This is good work, well done.

File: src/config/swagger.json (hunk 2 above)
Comment (daviddamilola) [reply src/config/swagger.json]: changes have been implemented.

File: .travis.yml (hunk 1 above)
Comment (daviddamilola) [reply .travis.yml]: changes have been implemented.

File: src/config/swagger.json (hunk 3 above)
Comment (daviddamilola) [reply src/config/swagger.json]: changes have been implemented

File: src/controllers/UserController.js (hunk 4 above)
Comment (daviddamilola) [reply src/controllers/UserController.js]: changes have been implemented

**Review by alatos2 — APPROVED** [review]

**Review by Naimaoye — APPROVED** [review]

File: src/routes/index.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import express from 'express';
```
Comment (hustlaviola) [inline src/routes/index.js]: This file has been deleted. Consider rebasing

File: src/tests/index.test.js [hunk 7]
```diff
@@ -18,13 +18,13 @@ describe('Handle requests on other endpoints', () => {
       });
   });
 
-  it('should return an error when a wrong url is provided', done => {
+  it('should return a welcome message for version 1 of the API', done => {
```
Comment (hustlaviola) [inline src/tests/index.test.js]: This test has been removed as well

File: src/config/swagger.json [hunk 8]
```diff
@@ -98,6 +98,66 @@
...
+            "name": "body",
+            "description": "Information needed to update the user password",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/updatePassword"
+            }
+          }
+        ],
+        "responses": {
+          "201": {
```
Comment (hustlaviola) [inline src/config/swagger.json]: You can change the status code here
```suggestion
          "200": {
```

File: src/controllers/UserController.js [hunk 9]
```diff
@@ -55,4 +56,62 @@ export default class UserController {
...
+      Responses.setSuccess(200, 'enter your new email');
+      return Responses.send(res);
+    }
+    const { password } = req.body;
+    const [, rowsAffected] = await UserService.resetPassword(password, data.email);
+    const {
+      id, email, createdAt, updatedAt
+    } = rowsAffected;
+    Responses.setSuccess(
+      201,
```
Comment (hustlaviola) [inline src/controllers/UserController.js]:
```suggestion
      200,
```

Comment (daviddamilola) [reply src/controllers/UserController.js]: Thanks for the feedback @hustlaviola , changes have been implemented.

**Comment by ldonjibson** [comment]
you still dey work on the stuff?

**Pull request merged**